src/MPAGSCipher/PlayfairCipher.cpp

```diff
@@ -0,0 +1,190 @@
+
+    // Remove duplicated letters
+    std::string l_encountered{""};
+
+    auto rem_dup = [&] (char l){
+        // check the char against a string of letters already found
+        std::size_t found = l_encountered.find(l);
+        // if char is in the list, mark for removal (return true)
+        if (found != std::string::npos){
+            return true;
+        // if char isn't already in list, add it to list and return false
+        }
+        else {
+            l_encountered += l;
+            return false;
+        }
+
+    };  
+
+    auto iter2 = std::remove_if(key_.begin(), key_.end(), rem_dup);
+    key_.erase(iter2,key_.end());
+
+    // Store the coords of each letter
+    // looping over each letter and storing coordinates and the letter in a map
+
+    // the position divided by 5 (as an integer) tells us the row number
+    // the position mod 5 tells us the column number
+   
+    for (std::size_t i{0}; i < keyLength_; i++){
+        std::size_t row_ind{i / gridSize_};
+        std::size_t col_ind{i % gridSize_};
+        
+        auto coords = std::make_pair(row_ind,col_ind);
+
+        ChartoCoordsMap_[key_[i]] = coords;
+        CoordstoCharMap_[coords] = key_[i];
+
+      
+    };
+
+    
+}
+
+std::string PlayfairCipher::applyCipher(const std::string& inputText, const CipherMode cipherMode) const {
+
+   
+
+    // Change J → I
+    auto jtoi = [] (char letter){
+        if (letter == 'J'){
+            letter = 'I';
+            return letter;
+        }
+        else {
+            return letter;
+        }
+
+    };
+
+    std::transform(inputText.begin(), inputText.end(), inputText.begin(), jtoi);
+
+    std::string examinedDigraphs{""};
+
+    // If repeated chars in a digraph add an X or Q if XX
+    // Look at each character in inputText
+    // If the character before is the same as the current one, replace with X + current character
+    // If both the characters are XX, replace with QX.
+    for (std::size_t i{0}; i < inputText.size(); i++){
+        // only doing something on every second letter
+        if (i == 1 % 2){
+            // checking if there are pairs of repeated letters
+            if (inputText[i] == inputText[i-1]){
+                // if the repeated letters are XX, populate examinedDigrahs with QX
+                if (inputText[i] == 'X'){
+                    examinedDigraphs+= "QX";
+                    
+                }
+                // if the repeated letters are not XX, populate examinedDigraphs with X(letter)
+                else{
+                    examinedDigraphs+= "X";
+                    examinedDigraphs+= inputText[i];
+                    
+
+                }
+            }
+            // if the pair are not the same letter, populate examinedDigraphs with the letters
+            else{
+                examinedDigraphs+= inputText[i-1];
+                examinedDigraphs+= inputText[i];
+                
+            }
```

Comment:
Now that we're looping in jumps of 2 and start by examining the `[0]` element, we need to do the following:
1. Always add the `[i]` element to the output.
2. Check that the `[i]` element is not the last in the input - if it is, add a 'Z' (or something else if element `[i]` is already a 'Z') and then break out of the loop
3. If `[i]` and `[i+1]` are different add `[i+1]` to the output
4. If they are the same, instead add 'X' to the output (or 'Q' if `[i]` is an 'X'), then decrement the loop counter so that the next iteration of the loop starts with the current `[i+1]` because we've not yet added that one to the output and it has now become the first in the next digraph rather than the second in this one